v2/protocol/pubsub/write_pubsub_message.go

```diff
@@ -0,0 +1,86 @@
+package pubsub
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/cloudevents/sdk-go/v2/binding"
+	"github.com/cloudevents/sdk-go/v2/binding/format"
+	"github.com/cloudevents/sdk-go/v2/binding/spec"
+	"github.com/cloudevents/sdk-go/v2/types"
+)
+
+// Fill the provided pubsubMessage with the message m.
+// Using context you can tweak the encoding processing (more details on binding.Write documentation).
+func WritePubSubMessage(ctx context.Context, m binding.Message, pubsubMessage *pubsub.Message, transformers ...binding.TransformerFactory) error {
+	structuredWriter := (*pubsubMessagePublisher)(pubsubMessage)
+	binaryWriter := (*pubsubMessagePublisher)(pubsubMessage)
+
+	_, err := binding.Write(
+		ctx,
+		m,
+		structuredWriter,
+		binaryWriter,
+		transformers...,
+	)
+	return err
+}
+
+type pubsubMessagePublisher pubsub.Message
+
+func (b *pubsubMessagePublisher) SetStructuredEvent(ctx context.Context, f format.Format, event io.Reader) error {
+	val, err := ioutil.ReadAll(event)
+	if err != nil {
+		return err
+	}
+	b.Data = val
+	return nil
+}
+
+func (b *pubsubMessagePublisher) Start(ctx context.Context) error {
+	b.Attributes = make(map[string]string)
+	return nil
+}
+
+func (b *pubsubMessagePublisher) End(ctx context.Context) error {
+	return nil
+}
+
+func (b *pubsubMessagePublisher) SetData(reader io.Reader) error {
+	val, err := ioutil.ReadAll(reader)
```

Comment:
Avoid `ioutil.ReadAll` when possible, it's really costly. Use `bytes.Buffer` + `io.Copy` which optimizes use cases when the initial reader was already a buffer: https://github.com/cloudevents/sdk-go/blob/master/v2/protocol/kafka_sarama/write_producer_message.go#L59

Reply:
done